src/transactions/FeeBumpTransactionFrame.cpp

```diff
@@ -0,0 +1,416 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "transactions/FeeBumpTransactionFrame.h"
+#include "crypto/Hex.h"
+#include "crypto/SHA.h"
+#include "crypto/SignerKey.h"
+#include "ledger/LedgerTxn.h"
+#include "ledger/LedgerTxnEntry.h"
+#include "ledger/LedgerTxnHeader.h"
+#include "main/Application.h"
+#include "transactions/SignatureChecker.h"
+#include "transactions/SignatureUtils.h"
+#include "transactions/TransactionUtils.h"
+#include "xdrpp/marshal.h"
+
+#include <numeric>
+
+namespace stellar
+{
+
+static TransactionEnvelope
+fromFeeBumpInnerTx(TransactionEnvelope const& envelope)
+{
+    TransactionEnvelope e(ENVELOPE_TYPE_TX);
+    e.v1() = envelope.feeBump().tx.innerTx.v1();
+    return e;
+}
+
+static void
+updateResult(TransactionResult& outerRes, TransactionResult const& innerRes)
+{
+    if (innerRes.result.code() == txSUCCESS)
+    {
+        outerRes.result.results() = innerRes.result.results();
+    }
+    else
+    {
+        outerRes.result.code(txINNER_FAILED);
+        auto& inner = outerRes.result.innerResult();
+        inner.feeCharged = 0;
+        inner.result.code(innerRes.result.code());
+        if (inner.result.code() == txFAILED)
+        {
+            inner.result.results() = innerRes.result.results();
+        }
+    }
+}
+
+FeeBumpTransactionFrame::FeeBumpTransactionFrame(
+    Hash const& networkID, TransactionEnvelope const& envelope)
+    : mEnvelope(envelope)
+    , mInnerTx(std::make_shared<TransactionFrame>(
+          networkID, fromFeeBumpInnerTx(envelope), false))
+    , mNetworkID(networkID)
+{
+}
+
+// TODO(jonjove): INTERNAL_ERROR protection?
+bool
+FeeBumpTransactionFrame::apply(Application& app, AbstractLedgerTxn& ltx,
+                               TransactionMeta& meta)
+{
+    LedgerTxn ltxTx(ltx);
+    SignatureChecker signatureChecker{
+        ltxTx.loadHeader().current().ledgerVersion, getContentsHash(),
+        mEnvelope.feeBump().signatures};
+    {
+        auto feeSource = stellar::loadAccount(ltxTx, getFeeSourceID());
+        if (feeSource)
+        {
+            // TODO(jonjove): This is only done to remove one-time signers
+            checkSignature(
+                signatureChecker, feeSource,
+                feeSource.current().data.account().thresholds[THRESHOLD_LOW]);
+        }
+    }
+    removeUsedOneTimeSignerKeys(signatureChecker, ltxTx);
+
+    auto& txChanges =
+        meta.v() == 1 ? meta.v1().txChanges : meta.v2().txChangesBefore;
+    txChanges = ltxTx.getChanges();
+    ltxTx.commit();
+
+    TransactionMeta innerMeta(2);
+    bool res = mInnerTx->apply(app, ltx, innerMeta);
+    auto& innerTxChanges = innerMeta.v2().txChangesBefore;
+    txChanges.insert(txChanges.end(), innerTxChanges.begin(),
```

Comment:
it looks like you're not propagating the rest of the meta (`operations` and `txChangesAfter`).

When you do it, make sure to avoid using 2x memory (move from one container to the other).

Reply:
(seems like you're missing some basic sanity check tests that the meta is there)